lang/gjs/src/gtk4/astalify.ts

```diff
@@ -1 +1,293 @@
-// TODO:
+import Astal from "gi://Astal?version=4.0"
+import Gtk from "gi://Gtk?version=4.0"
+import Gdk from "gi://Gdk?version=4.0"
+import GObject from "gi://GObject"
+import { execAsync } from "../process.js"
+import Variable from "../variable.js"
+import Binding, { kebabify, snakeify, type Connectable, type Subscribable } from "../binding.js"
+
+export function mergeBindings(array: any[]) {
+    function getValues(...args: any[]) {
+        let i = 0
+        return array.map(value => value instanceof Binding
+            ? args[i++]
+            : value,
+        )
+    }
+
+    const bindings = array.filter(i => i instanceof Binding)
+
+    if (bindings.length === 0)
+        return array
+
+    if (bindings.length === 1)
+        return bindings[0].as(getValues)
+
+    return Variable.derive(bindings, getValues)()
+}
+
+function setProp(obj: any, prop: string, value: any) {
+    try {
+        // the setter method has to be used because
+        // array like properties are not bound correctly as props
+        const setter = `set_${snakeify(prop)}`
+        if (typeof obj[setter] === "function")
+            return obj[setter](value)
+
+        return (obj[prop] = value)
+    }
+    catch (error) {
+        console.error(`could not set property "${prop}" on ${obj}:`, error)
+    }
+}
+
+export default function astalify<
+    C extends { new(...args: any[]): Gtk.Widget },
+>(cls: C) {
+    class Widget extends cls {
+        get css(): string { return Astal.widget_get_css(this) }
+        set css(css: string) { Astal.widget_set_css(this, css) }
+        get_css(): string { return this.css }
+        set_css(css: string) { this.css = css }
+
+        get className(): string { return Astal.widget_get_class_names(this).join(" ") }
+        set className(className: string) { Astal.widget_set_class_names(this, className.split(/\s+/)) }
+        get_class_name(): string { return this.className }
+        set_class_name(className: string) { this.className = className }
+
+        get cursor(): Cursor { return Astal.widget_get_cursor(this) as Cursor }
```

Comment:
`cursor` conflicts with `Gtk.Widget` cursor: this becomes type of `Gdk.Cursor & string`. maybe other name?